tiledb/cloud/dag/dag.py

```diff
@@ -792,3 +763,100 @@ def end_results_by_name(self):
             results[node.name] = node.result()
 
         return results
+
+
+class _ReplaceWith(Exception):
+    """Indicates that the object should be replaced rather than descending."""
+
+    def __init__(self, value: Any):
+        self.value = value
+
+
+def replace_stored_params(tree, loader: stored_params.ParamLoader) -> Any:
+    """Descends into data structures and replaces Stored Params with results.
+
+    This is only useful in the UDF execution environment.
+
+    Guaranteed to produce new output structures and not mutate the input.
+    Supports cyclic data structures.
+    """
+
+    def visit(arg):
+        if isinstance(arg, stored_params.StoredParam):
+            raise _ReplaceWith(loader.load(arg))
+
+    return _ResultReplacer(visit)._replace_internal(tree)
+
+
+def _replace_nodes_with_results(tree):
+    """Descends into data structures and replaces Node IDs with their values.
+
+    Offers the same guarantees as :func:`replace_stored_params`.
+    """
+
+    def visit(arg):
```

Comment:
In this case, `_replace_nodes_with_results` is only called once per graph; everything else happens inside the `_ResultReplacer`.

That said, looking at this now I do think this would be clearer with a slightly more traditional abstract class–based approach, so I’ve reworked it.